Adding benchmark results and a system overview to the README. (#294)
y-scope/clp: README.md

[![Open feature requests](https://img.shields.io/github/issues/y-scope/clp/enhancement?label=feature-requests)](https://github.com/y-scope/clp/issues?q=is%3Aissue+is%3Aopen+label%3Aenhancement)
[![CLP on Zulip](https://img.shields.io/badge/zulip-yscope--clp%20chat-1888FA?logo=zulip)](https://yscope-clp.zulipchat.com/) 

> YScope's Compressed Log Processor (CLP) compresses text logs, and allows you to search the compressed

Reviewer: ```suggestion
YScope's Compressed Log Processor (CLP) compresses your logs, and allows you to search the compressed
```
This is minor, but I think "text logs" may be misunderstood as "unstructured logs".

also supports real-time log compression within several logging libraries. CLP also includes 
purpose-built web interfaces for searching and viewing the compressed logs. 
To learn more about it, you can read our 
[paper](https://www.usenix.org/system/files/osdi21-rodrigues.pdf).

# Benchmarks

![CLP Benchmark on JSON Logs](docs/img/clp-json-benchmark.png)
![CLP Benchmark on Unstructured Logs](docs/img/clp-unstructured-benchmark.png)

The figures above show CLP's compression and search performance compared to other tools. We separate
the experiments between JSON and unstructured logs because (1) some tools can only handle one type of
logs, and (2) tools that can handle both types often have different designs for each type (such as CLP). 

Compression ratio is measured as the average across a variety of log datasets. Some of these datasets
can be found [here](docs/Datasets.md). Search performance is measured using queries on the MongoDB logs
(for JSON) and the Hadoop logs (for unstructured logs). Note that CLP uses an index-less design, so for
a fair comparison, we disabled MongoDB and PostgreSQL's indexes; If we left them enabled, MongoDB and

> PostgreSQL's compression ratio would be worse. We didn’t disable indexing for Elasticsearch or Splunk

Reviewer: ```suggestion
PostgreSQL's compression ratio would be worse. We didn't disable indexing for Elasticsearch or Splunk
```

> since these tools are fundamentally index-based tools (i.e., logs cannot be searched without indexes).

Reviewer: ```suggestion
since these tools are fundamentally index-based (i.e., logs cannot be searched without indexes).
```

More details about our experimental methodology can be found in the 
[CLP paper](https://www.usenix.org/system/files/osdi21-rodrigues.pdf).

# System Overview

![CLP systems overview](docs/img/clp-complete-solution.png)

CLP provides an end-to-end log management pipeline consisting of compression, search, analytics, and

> viewing. The figure above shows the system architecture. It consists of the following features:

Reviewer: ```suggestion
viewing. The figure above shows CLP's architecture. It consists of the following features:
```

Or perhaps "shows the CLP ecosystem."?

  [web UI](components/webui). The input can either be raw logs
  or CLP's compressed IR (intermediate representation) produced by CLP's logging libraries.

- **Real-time Compression with CLP Logging Libraries**: CLP provides logging libraries for 
  [Python](https://github.com/y-scope/clp-loglib-py) and Java ([Log4j](https://github.com/y-scope/log4j1-appenders) 
  and [Logback](https://github.com/y-scope/logback-appenders)). The logging libraries compress logs in
  real-time, so only compressed logs are written to disk or transmitted over the network. The compressed
  logs use CLP's intermediate representation (IR) format which achieves a higher compression ratio
  than general purpose compressors like Zstandard. Compressing IR into archives can further double the
  compression ratio and enable global search, but this requires more memory usage as it needs to buffer
  enough logs. More details on IR versus archives can be found in this [Uber Engineering Blog](https://www.uber.com/en-US/blog/reducing-logging-cost-by-two-orders-of-magnitude-using-clp).

- **[Log Viewer](https://github.com/y-scope/yscope-log-viewer)**: the compressed IR can be viewed in a 

>   webUI log viewer. Compared to viewing the logs in an editor, CLP's log viewer supports advanced 

Reviewer: ```suggestion
  web-based log viewer. Compared to viewing the logs in an editor, CLP's log viewer supports advanced 
```
